src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1365 @@
+                op,
+                self.allocs[(inst.index(), op_idx)]
+            );
+            return Ok(());
+        }
+        if !self.allocd_within_constraint(inst, op) {
+            trace!("{op} isn't allocated within constraints.");
+            let curr_alloc = self.vreg_allocs[op.vreg().vreg()];
+            let new_alloc = self.alloc_operand(inst, op, op_idx)?;
+            if curr_alloc.is_none() {
+                self.live_vregs.insert(op.vreg());
+                self.vreg_to_live_inst_range[op.vreg().vreg()].1 = match (op.pos(), op.kind()) {
+                    (OperandPos::Late, OperandKind::Use) | (_, OperandKind::Def) => {
+                        // Live range ends just before the early phase of the
+                        // next instruction.
+                        ProgPoint::before(Inst::new(inst.index() + 1))
+                    }
+                    (OperandPos::Early, OperandKind::Use) => {
+                        // Live range ends just before the late phase of the current instruction.
+                        ProgPoint::after(inst)
+                    }
+                };
+                self.vreg_to_live_inst_range[op.vreg().vreg()].2 = new_alloc;
+
+                trace!("Setting vreg_allocs[{op}] to {new_alloc:?}");
+                self.vreg_allocs[op.vreg().vreg()] = new_alloc;
+                if let Some(preg) = new_alloc.as_reg() {
+                    self.vreg_in_preg[preg.index()] = op.vreg();
+                }
+            }
+            // Need to insert a move to propagate flow from the current
+            // allocation to the subsequent places where the value was
+            // used (in `prev_alloc`, that is).
+            else if curr_alloc.is_some() {
+                trace!("Move reason: Prev allocation doesn't meet constraints");
+                if self.is_stack(new_alloc)
+                    && self.is_stack(curr_alloc)
+                    && self.edits.scratch_regs[op.class()].is_none()
+                {
+                    let reg = self.get_scratch_reg(op.class())?;
+                    self.edits.scratch_regs[op.class()] = Some(reg);
+                    self.available_pregs[OperandPos::Early].remove(reg);
+                    self.available_pregs[OperandPos::Late].remove(reg);
+                }
+                if op.kind() == OperandKind::Def {
+                    trace!("Adding edit from {new_alloc:?} to {curr_alloc:?} after inst {inst:?} for {op}");
+                    self.edits.add_move(
+                        inst,
+                        new_alloc,
+                        curr_alloc,
+                        op.class(),
+                        InstPosition::After,
+                    );
+                    // No need to set vreg_in_preg because it will be set during
+                    // `freealloc` if needed.
+                }
+                // Edits for use operands are added later to avoid inserting
+                // edits out of order.
```

Comment:
If the vreg doesn't live past the instruction then `curr_alloc` is `None` and no move needs to be emitted.

I'm talking about a case like this:
```
1. use v1 (fixed: p1)
2. use v1 (fixed: p0)
```

In this case `curr_alloc` will be `p0` and `new_alloc` will be `p1`. A move needs to be inserted from `new_alloc` to `curr_alloc` after the instruction.

Reply:
If there are multiple operands using the same vreg in different allocations in the same instruction, then the moves will have to be inserted before:
```
1. use v1 (fixed: p1), use v1 (fixed: p2), use v1 (fixed: p3)
```
When the second v1 operand is encountered, `curr_alloc` will be p1 and inserting a move after will be incorrect since v1 doesn't live past this instruction.

Reply:
Yes but that's a completely different kind of move that is handled in a different place (in the loop on line 948).

What I'm trying to say is that `if op.kind() == OperandKind::Def` on line 640 is incorrect, the moves need to be inserted for both Use and Def operands after the instruction. Multiple uses *also* need to have moves inserted before the instruction, but as I said before, that's handled elsewhere.

Reply:
When `curr_alloc` != `new_alloc` for use operands, it is always safe to insert the edit before the instruction. In some cases, it is safe to add the edit after, but not in all cases like when a vreg is used multiple times in an instruction and that vreg doesn't live past the instruction. For example:

```
1. def v0 (fixed: p1), use v1 (fixed: p1), use v1 (fixed: p2)
2. use v0 (fixed: p1)
```

In the above, when the second use operand is encountered in inst 1, v1's `curr_alloc` will be p1 and `new_alloc` will be p2. If we decide to insert an edit after, we'll have:


```
1. def v0 (fixed: p1), use v1 (fixed: p1), use v1 (fixed: p2)
move from p2 to p1
3. use v0 (fixed: p1)
```

Overwriting v0, resulting in an incorrect allocation. Checking for this special case of multiple operands in the same instruction will also be more expensive than simply checking if the operand is a def operand. The algorithm does not remember anything about whether or not the operand being allocated is the first or hundredth use of the same vreg.

On the other hand, if we decide to always insert edits for use operands before the instruction instead of after, there will be no incorrect allocations because it is always safe to insert the edits before the instruction for use operands. Remember that `vreg_allocs`, once initialized for a vreg, doesn't change unless the vreg is evicted or deallocated. So, if a use operand is expected to be in some allocation p0 after an instruction, like in:

```
1. use v1 (fixed: p1)
2. use v1 (fixed: p0)
```

When v1 is encountered in inst 2, `vreg_allocs` is set to p0 for it. When it's encountered in inst 1. `curr_alloc` will be p0 and `new_alloc` will be p1. `vreg_allocs` doesn't change and v1's `curr_alloc` remains p0. The loop afterwards that handles insertion of use operand edits will correctly insert the move from p0 to p1 before the instruction.

> What I'm trying to say is that if op.kind() == OperandKind::Def on line 640 is incorrect, the moves need to be inserted for both Use and Def operands after the instruction. Multiple uses also need to have moves inserted before the instruction, but as I said before, that's handled elsewhere.

No, it's not incorrect. Such a move definitely needs to be inserted after the instruction for a def, because the defined vreg doesn't live before the instruction. But it can always be safely inserted before the instruction for use operands. Can you give an example in which inserting before the instruction for a use operand will lead to an incorrect allocation?

Reply:
You are correct. You method is actually better than what I had in mind.